File: SourceCode/Server/services/booking-service/srs/repositories/seat.model.js
const mongoose = require("mongoose");

const seatSchema = new mongoose.Schema(
  {
    showtimeId: {
      type: mongoose.Schema.Types.ObjectId,
      ref: "Showtime",
      required: [true, "Showtime is required"],
    },
    seatNumber: {
      type: String,
      required: [true, "Seat number is required"],
    },
    row: {
      type: String,
      required: [true, "Row is required"],
    },
    type: {
      type: String,
      enum: ["regular", "vip", "couple"],
      default: "regular",
    },
    status: {
      type: String,
      enum: ["available", "reserved", "booked"],
      default: "available",
    },
    reservedBy: {
      type: mongoose.Schema.Types.ObjectId,
      ref: "User",
    },
    reservedUntil: {
      type: Date, // Ghế được giữ trong 10 phút
    },
  },
  {
    timestamps: true,
  }
);

// Index để tìm ghế nhanh
seatSchema.index({ showtimeId: 1, status: 1 });

module.exports = mongoose.model("Seat", seatSchema);

File: SourceCode/Client/lib/features/home/models/seat_model.dart
class Seat {
  final String id;
  final String showtimeId;
  final String seatNumber;
  final String row;
  final String type; // regular, vip, couple
  final String status; // available, reserved, booked
  final String? reservedBy;
  final DateTime? reservedUntil;

  Seat({
    required this.id,
    required this.showtimeId,
    required this.seatNumber,
    required this.row,
    required this.type,
    required this.status,
    this.reservedBy,
    this.reservedUntil,
  });

  factory Seat.fromJson(Map<String, dynamic> json) {
    return Seat(
      id: json['_id'] ?? '',
      showtimeId: json['showtimeId'] ?? '',
      seatNumber: json['seatNumber'] ?? '',
      row: json['row'] ?? '',
      type: json['type'] ?? 'regular',
      status: json['status'] ?? 'available',
      reservedBy: json['reservedBy'],
      reservedUntil: json['reservedUntil'] != null
          ? DateTime.parse(json['reservedUntil'])
          : null,
    );
  }

  Map<String, dynamic> toJson() {
    return {
      '_id': id,
      'showtimeId': showtimeId,
      'seatNumber': seatNumber,
      'row': row,
      'type': type,
      'status': status,
      'reservedBy': reservedBy,
      'reservedUntil': reservedUntil?.toIso8601String(),
    };
  }

  bool get isAvailable => status == 'available';
  bool get isReserved => status == 'reserved';
  bool get isBooked => status == 'booked';
}
